src/olympia/addons/views.py

```diff
@@ -812,33 +813,127 @@ def as_view(cls, **initkwargs):
         return non_atomic_requests(
             super(LanguageToolsView, cls).as_view(**initkwargs))
 
-    def get_application_id(self):
-        if not hasattr(self, 'application_id'):
+    def get_query_params(self):
+        """
+        Parse query parameters that this API supports:
+        - app (mandatory)
+        - type (optional)
+        - appversion (optional, makes type mandatory)
+
+        Can raise ParseError() in case a mandatory parameter is missing or a
+        parameter is invalid.
+
+        Returns a tuple with application, addon_types tuple (or None), and
+        appversions dict (or None) ready to be consumed by the get_queryset_*()
+        methods.
+        """
+        # app parameter is mandatory when calling this API.
+        try:
+            application = AddonAppQueryParam(self.request).get_value()
+        except ValueError:
+            raise exceptions.ParseError('Invalid or missing app parameter.')
+
+        # appversion parameter is optional.
+        if AddonAppVersionQueryParam.query_param in self.request.GET:
             try:
-                self.application_id = AddonAppQueryParam(
-                    self.request).get_value()
+                value = AddonAppVersionQueryParam(self.request).get_values()
+                appversions = {
+                    'min': value[1],
+                    'max': value[2]
+                }
             except ValueError:
-                raise exceptions.ParseError('Invalid app parameter.')
-        return self.application_id
+                raise exceptions.ParseError('Invalid appversion parameter.')
+        else:
+            appversions = None
+
+        # type is optional, unless appversion is set. That's because the way
+        # dicts and language packs have their compatibility info set in the
+        # database differs, so to make things simpler for us we force clients
+        # to filter by type if they want appversion filtering.
+        if AddonTypeQueryParam.query_param in self.request.GET or appversions:
```

Comment:
Okay, I guess this answers my question from documentation above, but is this just because it's easier or is there a reason?  I.e. can we return dictionaries too if no type is specified (they would always be the same dictionaries, regardless of app version, obviously)?

Reply:
The problem is that dictionaries are "always compatible", so their behaviour is the complete opposite to the one for language packs, which are strictly compatible with a specific app version. So we'd need to make the queryset a lot more complex to handle that... I might revisit this at some point, but for now I wanted to get this implementation out of the door and see if that was good enough.

Reply:
okay that's fair.